### DISCUSSION TITLE
question(discord.js): Discord.js Development Release

### DISCUSSION THREAD

**Discussion by PenPow** [PenPow/SignalBot#16] [Development]
**Describe the solution you'd like**
Right now, discord.js@dev is our current library, but is not currently considered stable, due to the fact that it is a development build. This leads to many issues in production, but as we are still in development this is currently fine.

Right now we are on track to launch the bot _after_ V13 releases, but should there be a major holdup from @discordjs **or** should we make good time, we can launch prior to this, and replace it with V13 once that officially launches.

This issue will be serving as a discussion point for whether, if possible, we do stick with  djs dev into production/launch before the stable release, and also serve as a list of bugs that need patching with the breaking changes V13 has created.

The issues and bugs that are created will be tracked here, and then implemented in a PR at a later date. Currently we are still awaiting proper and finalised documentation of changes with djs dev 

cc @KyomiDev for his opinion on whether we stick with djs dev into the future, to get the newest features, considering the majority of updates will not land on the stable branch for up to 6 months after creation and secured, finalised implementation into the client side end.

**Describe alternatives you've considered**
There are a couple of options for our implementation

1. Immediately Downgrade to latest stable (12.5.3) and wait for V13 stable, losing us access to slash commands, replies, threads, and more.
2. Remain on djs dev until V13 stable, and then jump ship, providing us support but losing access to newest features.
3. Remain on djs dev forever, getting access to the newest features, but almost daily breaking changes.

**Additional context**
For additional context regarding the update, djs dev, and V13.

The discord API lib/wrapper we use is discord.js, as it would make no sense to write raw API calls. Over the past year, djs has been working on releasing V13, which implements almost all of the new developer features released by discord since Version 12.

Version 12, is the stable branch, which we should be aiming to use in production, the issue is this lacks key features and functionalities of the bot. Currently, this branch is depreciated, and in a couple of months, no support will be given.

Version 13, is not released yet, but will be the next stable update, providing access to slash commands, replies, threads, guild avatars, and more. This is what I am aiming for for Signal to move onto once it releases.

djs dev is the development build, that we are currently using, think of it as a beta for Version 13. It provides all the latest updates, and features, but at the cost of no notice breaking changes. Over the past 3 months, this has included everything from changing the 2 main events (messageCreate, and interactionCreate), and completely breaking the send and edit function.

The reason why this is so decisive, is because djs has a sort of reputation for taking a while with getting releases out compared to other libraries. While this isn't necessarily a bad thing, as djs always has a great implementation of the features, and i have found little to no bugs with it in my 4 years of discord bot development, it has meant that changes that may be stable in other libraries a week after the docs from discord are finalised, discord.js is still only on a draft pull request, which means our decisions are impacted by how much we need the features.

Normally, even in a development build of a bot, I would be implementing the stable library. However, as the launch is now just dependant on discord to finalise everything, V13 and djs dev allows us to start work without having to rewrite the bot when we have immediately launched.

Finally, we have the main issue that a lot of the features in djs dev are really useful, that we, as bot developers want to get our hands on. For example, the 2 main updates with djs dev provide me with access to slash commands, which are amazing considering the bot/messages API is depreciated (cc @KyomiDev for a further look into removing messages from the bot and going interaction only, will allow you to do some testing with that), but also because it means that we can be more data conscious, and get access to ephemeral responses, which means less clutter for server owners, allowing them to allow commands in any channel.

The other feature that I am using heavily in Signal are replies, which take use of actual reply functionality, rather then prepending @user to the front of the message. I use this because it shows easily who called the message, more context, and looks cleaner than just sending a random embed, as it allows server owners to monitor the channel, and more easily punish users who violate their rules by spamming long embeds.

In short, we need djs dev for the updates, as it takes an incredibly long time for the updates to come out on stable. This issue is a thread to discuss whether we remain on djs dev, so when discord launches new features, we are able to take full advantage on them.

**Comment by KyomiDev** [comment]
Just removed the accidental pings of the person who unfortunately named themselves dev

**Reply by PenPow** [reply]
Oops, sorry

**Comment by KyomiDev** [comment]
Ok, starting off with the main topic of the thread, versions.

I think we stick with dev for the meantime, and then should there be a feature we really need, then we can assess the costs of moving to a development build, over the usefulness of the feature.

**Reply by PenPow** [reply]
Sounds good to me!

**Comment by KyomiDev** [comment]
Regarding point 2, whether we ditch regular commands, I feel that they are a core part of user interactions. I believe we continue maintaining messages until either

A) Discord drops support
B) Discord.js drops support
C) We know more about how permissions work with slash commands
D) The general discord public widely adopts them.

Until then, we continue maintaining messages.

Note that I still believe that slash commands and interactions are the way forward, as it eases the burden of complication on the user, and allows us as developers to make it easier for users to interact.

Im wanting to make a PR to add more buttons into the bot, as the only current usage is for moderation confirmation. (Also need to patch that function to work with interactions)

**Reply by PenPow** [reply]
> Until then, we continue maintaining messages.

Alright, ill do some local research into punishment permissions for bot's with only slash commands.

**Reply by PenPow** [reply]
Following up on this, as I have a Semver Major PR in the works, I am wanting to come back to this.

Discord currently has depreciated the bot scope, and are pushing people to move solely to interactions, and Discord.js is also pushing people to move away from it as well.

The general discord public is still moving, but I believe that we can start to move over as well.

I will be opening a semver major PR within the coming days that completely removes all message scopes, along with making it so that we are slash commands only.

I will also be opening a semver minor PR removing all unused code.

**Reply by KyomiDev** [reply]
LGTM

**Reply by KyomiDev** [reply]
> I will be opening a semver major PR within the coming days that completely removes all message scopes, along with making it so that we are slash commands only.

If you want to work on your current PR, I can get started with this one, continue to implement messages into it as this PR will take a fair share longer.

**Reply by KyomiDev** [reply]
> I will also be opening a semver minor PR removing all unused code.

Should I implement this into the PR referenced ^ or shall we put this in its own PR.

**Reply by PenPow** [reply]
> > I will also be opening a semver minor PR removing all unused code.
> 
> Should I implement this into the PR referenced ^ or shall we put this in its own PR.

I see no harm in putting it into that combined PR
